tools/src/main/java/org/apache/kafka/tools/PushHttpMetricsReporter.java

```diff
@@ -0,0 +1,317 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.tools;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.config.AbstractConfig;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.errors.InterruptException;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.UnknownHostException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * MetricsReporter that aggregates metrics data and reports it via HTTP requests to a configurable
+ * webhook endpoint in JSON format.
+ */
+public class PushHttpMetricsReporter implements MetricsReporter {
+    private static final Logger log = LoggerFactory.getLogger(PushHttpMetricsReporter.class);
+
+    private static final String METRICS_PREFIX = "metrics.";
+    static final String METRICS_URL_CONFIG = METRICS_PREFIX + "url";
+    static final String METRICS_PERIOD_CONFIG = METRICS_PREFIX + "period";
+    static final String METRICS_HOST_CONFIG = METRICS_PREFIX + "host";
+    static final String CLIENT_ID_CONFIG = ProducerConfig.CLIENT_ID_CONFIG;
+
+    private static final Map<String, String> HEADERS = new LinkedHashMap<>();
+    static {
+        HEADERS.put("Content-Type", "application/json");
+    }
+
+    private final Object lock = new Object();
+    private final Time time;
+    private final ScheduledExecutorService executor;
+    private URL url;
+    private String host;
+    private String clientId;
```

Comment:
Nit: can we move all non-final classes after the final ones? Also, we can maybe add a comment above the non-final ones stating that they are set via `configure`. It might also be worth mentioning that `metrics` gets updated via `init`/`metricChange`/`metricRemoval`.

That makes it easier for someone to understand the lifecycle without having to look at the whole class.

Reply:
Ack